sdk/dotnet/Pulumi.Automation/XStack.cs

```diff
@@ -0,0 +1,631 @@
+            var execKind = ExecKind.Local;
+            var program = this.Workspace.Program;
+            var args = new List<string>()
+            {
+                "up",
+                "--yes",
+                "--skip-preview",
+            };
+
+            if (options != null)
+            {
+                if (options.Program != null)
+                    program = options.Program;
+
+                if (!string.IsNullOrWhiteSpace(options.Message))
+                {
+                    args.Add("--message");
+                    args.Add(options.Message);
+                }
+
+                if (options.ExpectNoChanges is true)
+                    args.Add("--expect-no-changes");
+
+                if (options.Replace?.Any() == true)
+                {
+                    foreach (var item in options.Replace)
+                    {
+                        args.Add("--replace");
+                        args.Add(item);
+                    }
+                }
+
+                if (options.Target?.Any() == true)
+                {
+                    foreach (var item in options.Target)
+                    {
+                        args.Add("--target");
+                        args.Add(item);
+                    }
+                }
+
+                if (options.TargetDependents is true)
+                    args.Add("--target-dependents");
+
+                if (options.Parallel.HasValue)
+                {
+                    args.Add("--parallel");
+                    args.Add(options.Parallel.Value.ToString());
+                }
+            }
+
+            InlineLanguageHost? inlineHost = null;
+            try
+            {
+                if (program != null)
+                {
+                    execKind = ExecKind.Inline;
+                    inlineHost = new InlineLanguageHost(program, cancellationToken);
+                    await inlineHost.StartAsync().ConfigureAwait(false);
+                    var port = await inlineHost.GetPortAsync().ConfigureAwait(false);
+                    args.Add($"--client=127.0.0.1:{port}");
+                }
+
+                args.Add("--exec-kind");
+                args.Add(execKind);
+
+                var upResult = await this.RunCommandAsync(args, options?.OnOutput, cancellationToken).ConfigureAwait(false);
+                var output = await this.GetOutputAsync(cancellationToken).ConfigureAwait(false);
+                var summary = await this.GetInfoAsync(cancellationToken).ConfigureAwait(false);
+
+                return new UpResult(
+                    upResult.StandardOutput,
+                    upResult.StandardError,
+                    summary!,
+                    output);
+            }
+            finally
+            {
+                if (inlineHost != null)
+                {
+                    await inlineHost.DisposeAsync().ConfigureAwait(false);
```

Comment:
One thing to consider is that the language runtime server is running arbitrary user code that could throw, dereference null values, etc. 

Have we done any manual testing to ensure that the resulting error is caught and returned to the user?

Reply:
This is an excellent point and I don't think we are handling it properly. I will write a test for it and see what I can do.

Reply:
I have written a test for this. Exceptions thrown in the user provided code are now propagated up to the caller as I would expect as a consumer when calling `UpAsync` or `PreviewAsync`. I think that captures what you had in mind?

Reply:
Yes, I believe it does. Just wanted to make sure that errors are handled gracefully and propagated to the user, and that the gRPC server doesn't leak in the error case.